### PR TITLE
TRUNK-5220 Migrate PatientMergeAction

### PR DESCRIPTION
**Pull request by fanyui**
## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5220

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java [hunk 1]
```diff
@@ -574,10 +582,18 @@ public void mergePatients(Patient preferred, Patient notPreferred) throws APIExc
...
+			for (PatientMergeAction patientMergeAction : patientMergeActions) {
+				patientMergeAction.beforeMergingPatients(preferred, notPreferred);
+			}
+		}
+		
 		requireNoActiveOrderOfSameType(preferred,notPreferred);
 		PersonMergeLogData mergedData = new PersonMergeLogData();
 		mergeVisits(preferred, notPreferred, mergedData);
-		mergeEncounters(preferred, notPreferred, mergedData);
+		mergeEncounters(preferred, notPreferred, mergedData);//merge the remaining encounter that don't have visit
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Was this comment part of the ticket?

File: api/src/main/java/org/openmrs/api/PatientService.java [hunk 2]
```diff
@@ -844,4 +844,27 @@ public void saveCauseOfDeathObs(Patient patient, Date dateDied, Concept causeOfD
...
+	public void removePatientMergeAction(PatientMergeAction patientMergeAction);
+	
+	public void setPatientMergeActions(List<PatientMergeAction> patientMergeActions);
+	
+	/**
+	 * Gets the patientmergeactions
+	 * 
+	 * @return returns a list of registered patient merge actions
+	 */
+	public List<PatientMergeAction> getPatientMergeActions();
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/PatientService.java]: I wouldn't expose these getters/setters for merge actions, this should be internal logic to the service that should not be exposed.

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java (hunk 1 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Not actually i just thought i should make it descriptive. But i will remove it while fixing the reviews Wluyima provided

File: api/src/main/java/org/openmrs/api/PatientService.java [hunk 3]
```diff
@@ -844,4 +844,27 @@ public void saveCauseOfDeathObs(Patient patient, Date dateDied, Concept causeOfD
 	 * @throws PatientIdentifierTypeLockedException
 	 */
 	public void checkIfPatientIdentifierTypesAreLocked() throws PatientIdentifierTypeLockedException;
+	
+	/**
+	 * Adds actions to be performed before and after merging patients
+	 * 
+	 * @param patientMergeActions merge actions to be registered
+	 */
+	public void addPatientMergeAction(PatientMergeAction patientMergeAction);
```
[the remark this answers is not in the document]
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/PatientService.java]: Ok

File: api/src/main/java/org/openmrs/api/PatientService.java (hunk 2 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/PatientService.java]: I was sort of not finding a better way to write tests  for it using the patientservice interface so i added it. Thanks. I think i really can do better than that.

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java [hunk 4]
```diff
@@ -1636,4 +1726,25 @@ public void checkIfPatientIdentifierTypesAreLocked() {
 			throw new PatientIdentifierTypeLockedException();
 		}
 	}
+	
+	public void addPatientMergeAction(PatientMergeAction patientMergeAction) {
+		if (this.patientMergeActions == null) {
+			this.patientMergeActions = new ArrayList<PatientMergeAction>();
```
Comment (HelioStrike) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Why not put this

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java [hunk 5]
```diff
@@ -1636,4 +1726,25 @@ public void checkIfPatientIdentifierTypesAreLocked() {
...
+		}
+		this.patientMergeActions.add(patientMergeAction);
+	}
+	
+	@Override
+	public void removePatientMergeAction(PatientMergeAction patientMergeAction) {
+		if (this.patientMergeActions == null) {
+			this.patientMergeActions = new ArrayList<PatientMergeAction>();
```
Comment (HelioStrike) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: and this into a constructor? Or maybe the arraylist can be created at declaration.

**Comment by gracepotma** [comment]
@dkayiwa to review - can this be closed?

**Comment by dkayiwa** [comment]
Yes closing this because it was stolen by @gitcliff 😊

**Pull request closed without merging**